jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/stats/TimeBasedSlidingWindow.kt

```diff
@@ -22,38 +22,42 @@ import java.time.Instant
 
 open class TimeBasedSlidingWindow<T : Any>(
     private val windowSize: Duration,
-    private val clock: Clock = Clock.systemUTC(),
-    private val evicted: (TimeEntry<T>) -> Unit = {}
-) : EvictingQueue<T, TimeBasedSlidingWindow.TimeEntry<T>>(
-    { value -> TimeEntry(clock.instant(), value) },
-    { entry -> Duration.between(entry.insertionTime, clock.instant()) > windowSize }
+    private val evictionHandler: (T) -> Unit = {},
+    private val clock: Clock = Clock.systemUTC()
 ) {
-    class TimeEntry<T : Any>(
-        val insertionTime: Instant,
-        override val value: T
-    ) : EvictingQueue.Entry<T> {
-        override fun equals(other: Any?): Boolean {
-            if (this.value === other) {
-                return true
-            }
-            if (other?.javaClass != value.javaClass) {
-                return false
-            }
-            other as T
-            return other == value
+    private val queue = java.util.ArrayDeque<TimeEntry>()
+
+    private val evictionPredicate: (TimeEntry) -> Boolean = { entry ->
+        Duration.between(entry.insertionTime, clock.instant()) > windowSize
+    }
+
+    fun add(value: T) {
+        queue.addFirst(TimeEntry(value))
+        evict()
+    }
+
+    fun values(): Collection<T> = queue.reversed().map { it.value }
+
+    private fun evict() {
+        while (evictionPredicate(queue.last)) {
```

Comment:
This throws `NoSuchElementException` if the queue is empty, doesn't it?

(I think the only way you can get an empty queue is if `evict()` is still running more than `windowSize` time after `add`, but that could happen if someone instantiates this with a small enough `windowSize` or your machine gets a bad enough scheduling hit.)

Some use of `queue.peekLast()` is probably the right answer, but I don't want to try to get the Kotlin syntax right in a GitHub comment box.

It might be worth having a test that a sliding window of duration 0 works, and always has the value 0.

Reply:
Don't know if you saw it in later commits, but this was fixed by checking if the queue is empty in the while condition.